### DISCUSSION TITLE
Safe to mutate `params` in middlewares ?

### DISCUSSION THREAD

**Discussion by franky47** [prisma/prisma#10446] [Q&A] [answered]
Hello,

I'm working on a [middleware](https://github.com/47ng/prisma-field-encryption) that adds transparent field-level encryption capabilities. The idea is the following:

- In write queries, encrypt targeted fields of `params.args.data` before `await next(params)`
- In most queries, decrypt fields that have an encrypted value in the result before returning

My question is the following: while it seems fairly safe to mutate the result (it comes from the database or subsequent middlewares and no user code has any reference to its contents), is it safe to mutate the incoming params _(ie: encrypt in-place)_ ? 

Could this pose problems in cases like this ?
```ts
const defaultUserData = {
  name: 'should not be mutated'
}

prisma.user.update({
  where: { id: 42 },
  data: defaultUserData
})
```

Thanks !

**Comment by aqrln** [comment, marked as answer]
Hi @franky47!

> is it safe to mutate the incoming params (ie: encrypt in-place) ?
>
> Could this pose problems in cases like this ?

I don't think it may pose problems for Prisma per se, but for the application code — yes, this may have unintended consequences and will be unsafe in cases like this, if objects are shared between multiple operations.

**Reply by franky47** [reply]
Thank you, I'll do a deep-clone of the `params` before mutating anything then.

**Reply by aqrln** [reply]
Maybe something like https://github.com/immerjs/immer will help to minimize the number of nested copies (if the overhead won't be more than it's worth).

**Reply by franky47** [reply]
So far I'm using [rfdc](https://www.npmjs.com/package/rfdc), I'll have to run some benchmarks against immer to see what's the fastest (obviously query performance is a big factor).

**Reply by franky47** [reply]
After some experiments, I found this line that could be problematic:
https://github.com/prisma/prisma/blob/24f152008faa5b97f6a982cdbd035f14f139aa39/packages/client/src/runtime/getPrismaClient.ts#L1085

Namely, creating a copy (whether a full clone with rfdc or a partial one with immer) does not result in updates to `params.args.data` being reflected in the database, due to the lack of deep-merging, and `changedParams` not being merged there either.

I'm going to look more closely and open a minimal reproduction repository. Ideally, we should be able to:
- Mutate `params.args.data` in middlewares
- Have the updated values sent to the database
- Be able to do so with immer or a deep-clone to avoid mutating user references to the data.

One possible way to ensure immutability of user references would be for Prisma to do a deep copy of the input arguments before calling middlewares (this copy can be skiped if there's none). This way middlewares work on a sandboxed object they can safely mutate without additional code.

**Reply by franky47** [reply]
Related: issue [#9522](https://github.com/prisma/prisma/issues/9522) and PR [#9523](https://github.com/prisma/prisma/pull/9523)

**Reply by aqrln** [reply]
Oh, wow, got it. Thanks for catching this! I added some labels to that PR so that we could triage it later. We should really fix this 💯

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
